phoenix-core/src/main/java/org/apache/phoenix/iterate/ExplainTable.java

```diff
@@ -284,6 +355,106 @@ protected void explain(String prefix, List<String> planSteps,
+            ExplainPlanAttributesBuilder explainPlanAttributesBuilder,
+            List<List<Scan>> scansList) {
+        try {
+            StringBuilder buf = new StringBuilder().append(REGION_LOCATIONS);
+            Set<RegionBoundary> regionBoundaries = new HashSet<>();
+            List<HRegionLocation> regionLocations = new ArrayList<>();
+            for (List<Scan> scans : scansList) {
+                for (Scan eachScan : scans) {
+                    getRegionsInRange(tableRef.getTable().getPhysicalName().getBytes(),
+                            eachScan.getStartRow(),
+                            eachScan.getStopRow(),
+                            true,
+                            false,
+                            regionBoundaries,
+                            regionLocations);
+                }
+            }
+            int maxLimitRegionLoc = context.getConnection().getQueryServices().getConfiguration()
+                    .getInt(QueryServices.MAX_REGION_LOCATIONS_SIZE_EXPLAIN_PLAN,
+                            QueryServicesOptions.DEFAULT_MAX_REGION_LOCATIONS_SIZE_EXPLAIN_PLAN);
+            if (explainPlanAttributesBuilder != null) {
+                explainPlanAttributesBuilder.setRegionLocations(
+                        Collections.unmodifiableList(regionLocations));
+            }
+            if (regionLocations.size() > maxLimitRegionLoc) {
+                int originalSize = regionLocations.size();
+                List<HRegionLocation> trimmedRegionLocations =
+                        regionLocations.subList(0, maxLimitRegionLoc);
+                buf.append(trimmedRegionLocations);
+                buf.append("...total size = ");
+                buf.append(originalSize);
+            } else {
+                buf.append(regionLocations);
+            }
+            buf.append(") ");
+            return buf.toString();
+        } catch (IOException | SQLException | UnsupportedOperationException e) {
+            LOGGER.error("Explain table unable to add region locations.", e);
+            return "";
+        }
+    }
+
+    /**
+     * Region boundary class with start and end key of the region.
+     */
+    private static class RegionBoundary {
+        private final byte[] startKey;
+        private final byte[] endKey;
+
+        RegionBoundary(byte[] startKey, byte[] endKey) {
+            this.startKey = startKey;
+            this.endKey = endKey;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            RegionBoundary that = (RegionBoundary) o;
+            return Bytes.compareTo(startKey, that.startKey) == 0
+                    && Bytes.compareTo(endKey, that.endKey) == 0;
+        }
+
+        @Override
+        public int hashCode() {
+            return 0;
```

Comment:
why hashcode of zero?

Reply:
since byte[] can't be compared for hashcode (as they don't have `hashcode()` impl), i had to use unique value.
if we were doing string based comparison, then i would have used string's hashcode and equals implementations directly.